File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/ui/episodes/EpisodePagerAdapter.java
package com.battlelancer.seriesguide.ui.episodes;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import com.battlelancer.seriesguide.util.TextTools;
import java.util.ArrayList;

class EpisodePagerAdapter extends FragmentStatePagerAdapter {

    @NonNull private final ArrayList<Episode> episodes;
    private final Context context;
    private final boolean isMultiPane;

    EpisodePagerAdapter(Context context, FragmentManager fm,
            @NonNull ArrayList<Episode> episodes, boolean isMultiPane) {
        super(fm);
        this.context = context;
        this.episodes = episodes;
        this.isMultiPane = isMultiPane;
    }

    @Override
    public int getCount() {
        return episodes.size();
    }

    @Override
    public Fragment getItem(int position) {
        return EpisodeDetailsFragment.newInstance(episodes.get(position).episodeId, isMultiPane);
    }

    @Override
    public int getItemPosition(@NonNull Object object) {
        /*
         * This breaks the FragmentStatePagerAdapter (see
         * http://code.google.com/p/android/issues/detail?id=37990), so we
         * just destroy everything!
         */
        // EpisodeDetailsFragment fragment = (EpisodeDetailsFragment)
        // object;
        // int episodeId = fragment.getEpisodeId();
        // for (int i = 0; i < mEpisodes.size(); i++) {
        // if (episodeId == mEpisodes.get(i).episodeId) {
        // return i;
        // }
        // }
        return POSITION_NONE;
    }

    @Nullable
    Integer getItemEpisodeTvdbId(int position) {
        if (position < episodes.size()) {
            return episodes.get(position).episodeId;
        } else {
            return null;
        }
    }

    @Override
    public CharSequence getPageTitle(int position) {
        Episode episode = episodes.get(position);
        return TextTools.getEpisodeNumber(context, episode.seasonNumber, episode.episodeNumber);
    }

    void updateEpisodeList(@NonNull ArrayList<Episode> list) {
        episodes.clear();
        episodes.addAll(list);
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/shows/episodes/EpisodePagerAdapter.kt
// Copyright 2023 Uwe Trottmann
// SPDX-License-Identifier: Apache-2.0

package com.battlelancer.seriesguide.shows.episodes

import android.annotation.SuppressLint
import android.content.Context
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import com.battlelancer.seriesguide.shows.database.SgEpisode2Numbers
import com.battlelancer.seriesguide.util.TextTools

/**
 * Maps [Episode] objects to [EpisodeDetailsFragment] pages.
 */
@Suppress("DEPRECATION") // ViewPager2 paging performance is bad, don't need its improvements
@SuppressLint("WrongConstant") // Behavior flag not recognized as valid.
internal class EpisodePagerAdapter(
    private val context: Context,
    fm: FragmentManager
) : FragmentStatePagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

    private val episodes = ArrayList<SgEpisode2Numbers>()

    override fun getCount(): Int {
        return episodes.size
    }

    override fun getItem(position: Int): Fragment {
        return EpisodeDetailsFragment.newInstance(episodes[position].id)
    }

    override fun getItemPosition(`object`: Any): Int {
        /*
         * This breaks the FragmentStatePagerAdapter
         * (see https://issuetracker.google.com/issues/36956111),
         * so just destroy everything!
         * Note: This might be fixed with ViewPager2.
         */
        // EpisodeDetailsFragment fragment = (EpisodeDetailsFragment)
        // object;
        // int episodeId = fragment.getEpisodeId();
        // for (int i = 0; i < mEpisodes.size(); i++) {
        // if (episodeId == mEpisodes.get(i).episodeId) {
        // return i;
        // }
        // }
        return POSITION_NONE
    }

    fun getItemEpisodeId(position: Int): Long? {
        return if (position < episodes.size) {
            episodes[position].id
        } else {
            null
        }
    }

    override fun getPageTitle(position: Int): CharSequence? {
        val episode = episodes[position]
        return TextTools.getEpisodeNumber(context, episode.season, episode.episodenumber)
    }

    fun updateEpisodeList(list: List<SgEpisode2Numbers>) {
        episodes.clear()
        episodes.addAll(list)
        notifyDataSetChanged()
    }
}
